covariance/covarianceBase.h

```diff
@@ -34,96 +39,139 @@ class covarianceBase {
+  void throwParCurr(const double mag = 1.);
   void throwParameters();
-  virtual void throwNominal(bool nomValues=true, int seed = 0);
+  void throwNominal(bool nomValues=false, int seed = 0);
   // Randomly throw the parameters in their 1 sigma range
   void RandomConfiguration();
   
   // Getters
   TMatrixDSym *getCovMatrix() { return covMatrix; };
   TMatrixDSym *getInvCovMatrix() { return invCovMatrix; };
-  bool getEvalLikelihood(int i) { return fParEvalLikelihood[i]; };
-  bool getDoStep(int i) { return fParDoStep[i]; };
+  bool getEvalLikelihood(const int i) { return _fFlatPrior[i]; };
 
-  virtual double getLikelihood();
+  virtual int CheckBounds();
+  double CalcLikelihood();
+  virtual double GetLikelihood();
 
   const char *getName() { return matrixName; };
-  virtual const char* getParName(int i) const { 
-    return fParNames[i];
+  // ETA - Why is this virtual?
+  virtual const char* getParName(const int i) const {
+    return _fNames[i].c_str();
   };
   std::string const getInputFile() const { return inputFile; };
 
   // Get diagonal error for ith parameter
-  double getDiagonalError(const int i) { 
+  const double getDiagonalError(const int i) { 
     return sqrt((*covMatrix)(i,i));
   }
 
+
+  // Adaptive Step Tuning Stuff
+  void resetIndivStepScale();
+  void useSeparateThrowMatrix(TString throwMatrixName, TString throwMatrixFile, TString parameterMeansName="");
+  void useSeparateThrowMatrix();
+  void saveAdaptiveToFile(TString outFileName, TString systematicName);
+
+  void setThrowMatrix(TMatrixDSym *cov);
+  void updateThrowMatrix(TMatrixDSym *cov);
+  void setNumberOfSteps(const int nsteps){
+    total_steps=nsteps;
+    if(total_steps>=lower_adapt)resetIndivStepScale();
+  }
+  // Set thresholds for MCMC steps
+  void setAdaptiveThresholds(int low_threshold=10000, int up_threshold=1000000){lower_adapt=low_threshold; upper_adapt=up_threshold;}
+
+  TMatrixDSym *getThrowMatrix(){return throwMatrix;}
+  TMatrixD *getThrowMatrix_CholDecomp(){return throwMatrix_CholDecomp;}
+  std::vector<double> getParameterMeans(){return par_means;}
+
   //========
   //DB Pointer return
+  //ETA - This might be a bit squiffy? If the vector gots moved from say a
+  //push_back then the pointer is no longer valid... maybe need a better 
+  //way to deal with this? It was fine before when the return was to an 
+  //element of a new array. There must be a clever C++ way to be careful
   //========
-  const double* retPointer(int iParam) {return &fParProp[iParam];}
+  const double* retPointer(int iParam) {return &(_fPropVal.data()[iParam]);}
 
-  const std::vector<double> & getNominalArray() const { return nominal; };
+  virtual std::vector<double> getNominalArray();
   const std::vector<double> getProposed() const;
-  double getNominal(int i) { 
-    return nominal[i]; };
-  double getParProp(int i) { 
-    return fParProp[i]; 
+  const double getParProp(const int i) {
+    return _fPropVal[i]; 
+  };
```

Comment:
Could you check if Henry's updated to covariance in this PR 
https://github.com/t2k-software/MaCh3/pull/548/files
are included, otherwise something may not work with T2K

Reply:
I don't see anything that is likely to cause a problem. I think we can sort this out on when we're adding in all of Henry's adaptive stuff.